Google.Api.Gax.Grpc/HeaderParameterExtractor.cs

```diff
@@ -13,54 +13,89 @@
+    /// This class is immutable.
     /// </summary>
     public sealed class HeaderParameterExtractor<TRequest>
     {
         private readonly IReadOnlyList<ParamExtraction> _extractions;
+        private readonly IReadOnlyDictionary<string, IEnumerable<ParamExtraction>> _groupings;
 
         /// <summary>
-        /// 
+        /// Create a new HeaderParameterExtractor
         /// </summary>
-        /// <param name="paramName"></param>
-        /// <param name="regex"></param>
-        /// <param name="selector"></param>
-        public HeaderParameterExtractor(string paramName, Regex regex, Func<TRequest, string> selector)
+        public HeaderParameterExtractor()
         {
-            _extractions = new List<ParamExtraction> { new ParamExtraction(paramName, regex, selector) };
+            _extractions = new List<ParamExtraction>();
         }
 
         private HeaderParameterExtractor(IEnumerable<ParamExtraction> extractions)
         {
             _extractions = new List<ParamExtraction>(extractions);
+
+            // It is important that we Reverse the parameters within the group.
+            // The extractions follow the `last successfully matched wins` rule for
+            // conflict resolution when there are multiple extractions for the same parameter name.
+            // (see `google/api/routing.proto` for further details)
+            // But we would prefer to stop evaluating on the first successfully matched extraction,
+            // thus the reverse.
+            _groupings = _extractions.GroupBy(pe => pe.ParamName)
+                .ToDictionary(paramGroup => paramGroup.Key,
+                    paramGroup => paramGroup.Reverse());
         }
 
         /// <summary>
-        /// 
+        /// Adds a new header parameter extraction instruction.
+        /// The extractions follow the `last successfully matched wins` rule for
+        /// conflict resolution when there are multiple extractions for the same parameter name.
+        /// (see `google/api/routing.proto` for further details)
         /// </summary>
-        /// <param name="paramName"></param>
-        /// <param name="regex"></param>
-        /// <param name="selector"></param>
+        /// <param name="paramName">The name of the parameter in the routing header.</param>
+        /// <param name="regexStr">The regular expression (in the string form) used to extract the value of the parameter.
+        /// Should have exactly one named capturing group.</param>
+        /// <param name="selector">A function to call on each request, to determine the string to extract the header value from.
+        /// The parameter must not be null, but may return null.</param>
         /// <returns></returns>
-        public HeaderParameterExtractor<TRequest> Add(string paramName, Regex regex, Func<TRequest, string> selector)
+        public HeaderParameterExtractor<TRequest> WithParameter(string paramName, string regexStr, Func<TRequest, string> selector)
```

Comment:
It's still like the parameter name to be inferred from the regex - I think that'll simplify calling code. Again, I'm happy to do that myself, but it'll need to be before the generator change is merged.

Reply:
We can't have parameter name directly match regex group name, because we can have parameters with `.` in their name (actual case from Spanner), just like in the implicit headers. I can add an override of `WithParameter` that will work for the majority of the cases.

Reply:
Is that not a valid capturing group name? Okay... in that case, I'd argue we should move in the opposite direction: use an unnamed capturing group instead. (It still deduplicates, just in the opposite direction.)